rpexpose.1
.TH RPEXPOSE 1 2008-02-01
.SH NAME
RpExpose \- The Ratpoison Graphical Pager
.SH SYNOPSIS
.br
.B rpexpose \-\-help 
| 
.B \-\-version
.br
.B rpexpose \-\-generate
.BI [ file ]
.br
.B rpespose \-\-select
.BI [ file ]
.br
.B rpexpose \-\-clean
.SH EXAMPLES
From
.B rpthumb
.br
.B ratpoison
\-c \*(lqfdump\*(rq |
.br
.B tr
\*(lq,:\*(rq \*(lq\\n\\n\*(rq |
.br
.B sed
\*(lqs/[()]//g\*(rq | 
.br
.B awk
\'/window/ {print $2}\' |
.br
.B rpexpose
\-\-generate \-

From 
.B rpselect
.br
.B ratpoison 
\-c "windows %s:%i:%n:%t" |
.br
.B sed 
\-e \*(lqs/^*:/1:/\*(rq \-e \*(lqs/^[+\-]:/0:/\*(rq |
.br
.B rpexpose 
\-\-select \-

Add the following lines to your
.B .ratpoionrc
file
.br
.B addhook
switchwin exec
.B rpthumb
.br
.B addhook
quit exec
.BR rpexpose " --clean
.br
.B bind
.BI [ key ]
.RB exec " rpselect
.SH DESCRIPTION
RpExpose is a graphical pager written for the Ratpoison window manager, but designed to be used with any window manager.
.SH OPTIONS
.TP
.B \-h \-\-help
Show summary of options
.TP
.B \-v \-\-version
Show the version of program
.TP
.B \-g \-\-generate \fIfile\fP
Generates a thumbnail image of the top level windows listed in \fIfile\fP.  \fIfile\fP should list the x window id\'s of all of the top level windows.  If \fIfile\fP is \-, then rpexpose will read from stdin.  Assuming that you are using Ratpoison, \fBrpthumb\fP (provided with \fBrpexpose\fP) is a prewritten wrapper around \fB\-\-generate\fP, and you should simply call that instead.
.TP
.B \-s \-\-select \fIfile\fP
Creates a dialog box with thumbnails of all of the open windows and allows you to select one to raise.  \fIfile\fP contains a list of the current windows the syntax of which is shown below in \fBFILES\fP.  Assuming that you are using Ratpoison, \fBrpselect\fP (provided with \fBrpexpose\fP) is a prewritten wrapper around \fB\-\-select\fP, and you should simply call that instead.
.TP
.B \-c \-\-clean
Cleans all of the image files in the \fI~/.rpexpose/\fP directory.  This function should be called when your window manager exits to ensure that the \fI~/.rpexpose/\fP directory does not fill up with garbage data.
.SH FILES
.TP
.B \-s \-\-select \fIfile\fP
Each line of \fIfile\fP must follow the following format: \fB[\fI1_if_selected,_0_otherwise\fP]:[\fIX_Window_ID\fP]:[\fIInternal_ID\fP]:[\fIWindow_Name\fP]\fP
.TP
.B ~/.rpexposerc
A sample rpexposerc is included along side rpexpose for your convenience.  The rpexposerc file is configured in a similar manner to Ratpoison's rc file.  The commands are described in the following section.
.SH RPEXPOSERC FILE
.TP
.B left (h j) | right (l k)
Select the left or right thumbnail
.TP
.B quit (q)
Quits \fBrpexpose\fP
.TP
.B select (Space Return)
Raises the current window and exits \fBrpexpose\fP
.TP
.B bind [\fIkey\fP] [\fIcommand\fP] 
Binds the \fIkey\fP to the given \fIcommand\fP.  Execute command every time \fIkey\fP is pressed
.TP
.B unbind [\fIkey\fP] 
Unbind the \fIkey\fP.  Remove the action associated with \fIkey\fP
.TP
.B exec [\fIcommand\fP]
Execute the supplied \fIcommand\fP
.TP
.B set [\fIoption\fP] [\fIvalue\fP]
Set the internal \fIoption\fP to the supplied \fIvalue\fP.  A list of options is supplied in the following section
.SH RPEXPOSE OPTIONS
.TP
.B widescreen [\fIboolean\fP]
\fBrpexpose\fP attempts to make the selection window fit the shape of the screen as closely as possible.  If you have a wide-screen monitor, set this to be true
.TP
.B thumb_padding [\fIinteger\fP]
\fBthumb_padding\fP is the space between each thumbnail, as well as the thumbnail and the edge of the window.
.TP
.B text_padding [\fIinteger\fP]
\fBtext_padding\fB is the whitespace padding around each of the numbers in the top left hand corner of the thumbnails.
.TP
.B border_padding [\fIinteger\fP]
\fBborder_padding\fP is the width of the border around the thumbnail of the selected window.
.TP
.B select_exec [\fIstring\fP]
\fBselect_exec\fP is set to the string to be executed in order to raise a window.  The values \fI%x\fP, \fI%n\fP, and \fI%i\fP are replaced with the \fIX_Window_ID\fP, \fIWindow_Name\fP, and \fIWindow_ID\fP respectively.


rpexpose.c
/*
 * Copyright (c) 2008, Nathan Lawrence
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the <organization> nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY <copyright holder> ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL <copyright holder> BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "rpexpose.h"

extern const int BUFFER_SIZE;

const char *argp_program_version = "rpexpose 0.1.19";

const char *argp_program_bug_address = "<nlawren2@uiuc.edu>";

static char doc[] = "";

static char args_doc[] = "";

static struct argp_option cmdline_options[] = {
	{"clean",		'c', NULL,		0, "Cleans the /var/run/rpexpose directory" },
	{"generate",	'g', "FILE",	0, "Generates ximage dumps from a list of top level windows"},
	{"select",		's', "FILE",	0, "Creates a selection dialog for a list of managed windows"},
	{0}
};

int open_file(char *filename){
	if( strcmp((g.file.name=filename),"-") )
		g.file.handle=fopen(g.file.name,"r");
	else
		g.file.handle=stdin;
}

static error_t parse_opt(int key, char *arg, struct argp_state *state){
	switch(key){
	case 'c':
		g.action=A_CLEAR;
		break;
	case 'g':
		g.action=A_GENERATE;
		open_file(arg);
		break;
	case 's':
		g.action=A_SELECT;
		open_file(arg);
		break;
	case ARGP_KEY_END:
		if( !g.action ){
			argp_usage(state);
			exit(1);
		}
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

static struct argp argp = { cmdline_options, parse_opt, args_doc, doc };

int main(int argc, char **argv){
	g.action = A_UNKNOWN;
	g.file.name = NULL;
	g.file.home = getenv("HOME");

	argp_parse(&argp, argc, argv, 0, 0, NULL);
	
	switch(g.action){
	case A_CLEAR:
		rpexpose_clean();
		break;
	case A_GENERATE:
		rpexpose_generate();
		break;
	case A_SELECT:
		rpexpose_select();
		break; 
	}
	
	fclose(g.file.handle);
	return 0;
}

int rpexpose_clean(){
	struct dirent *ep;
	
	char filename[BUFFER_SIZE];
	sprintf(filename,"%s/.rpexpose",g.file.home);

	DIR *directory = opendir(filename);

	assert( !directory, "Couldn't open $HOME/.rpexpose/\n");
	while( ep = readdir(directory) ){
		if( *(ep->d_name)=='.' ) continue;
		if( !strcmp(ep->d_name, "default") ) continue;
		sprintf(filename,"%s/.rpexpose/%s",g.file.home,ep->d_name);
		if( remove(filename) ){
			error("Could not delete file");
			closedir(directory);
			return 1;
		}
	}
	
	closedir(directory);
	return 0;
}

int rpexpose_generate(){
	g.x.display = XOpenDisplay(getenv("DISPLAY"));

	Window window;
	
	char filename[BUFFER_SIZE];

	sprintf(filename,"%s/.rpexpose/",g.file.home);

	DIR *directory = opendir(filename);
	if(!directory){
		assert( mkdir(filename,0755), "Could not create directory" );
	}else{
		closedir(directory);
	}

	while(!feof(g.file.handle)){
		fscanf(g.file.handle,"%i\n",&window);
		XImage *thumbnail=thumbnail_generate(window);

		sprintf(filename, "%s/.rpexpose/%i", g.file.home, window);
		thumbnail_write(thumbnail,filename);
		XDestroyImage(thumbnail);
	}

	XCloseDisplay(g.x.display);
	return 0;
}

int rpexpose_select(){
	g.status=S_STARTUP;
	g.x.display = XOpenDisplay(getenv("DISPLAY"));

	load_rcdefaults();
	load_rcfile();
	load_input();
	load_xwindow();
	load_buffer();

	atexit(clean_up);

	// And here is when the magic is supposed to happen...
	
	XMapWindow(g.x.display,g.x.window);

	g.status=S_RUNNING;

	XEvent e;
	for(;;){
		XNextEvent(g.x.display, &e);
		switch(e.type){
		case MapNotify:
			XGrabKeyboard(g.x.display,g.x.window,False,GrabModeSync,GrabModeAsync,CurrentTime);
			break;
		case KeyPress:{
			KeySym keysym=XKeycodeToKeysym(g.x.display,e.xkey.keycode,0);
			if( keysym<256 && isdigit(keysym) ){
				if( g.p.selected->children[keysym-'0'] ){
					g.p.selected=g.p.selected->children[keysym-'0'];
					event_move(g.p.selected->window);
				}
				break;
			}
			KeyCode keycode=XKeysymToKeycode(g.x.display,keysym);
			parse_command(g.rc.keybindings[keycode]);
			break;
		}
		case Expose:
			event_redraw(e.xexpose.x,e.xexpose.y,e.xexpose.width,e.xexpose.height);
			event_move(g.gui.selected);
		}
	}
	return 0;
}


void clean_up(){
	g.status=S_SHUTDOWN;

	XUngrabKeyboard(g.x.display,CurrentTime);
	XUnmapWindow(g.x.display,g.x.window);
	
	{
		int i;
		for(i=0; i<256; ++i) free(g.rc.keybindings[i]);
	}

	{
		thumbnail_t *i=g.gui.thumbs, *prev;
		i->right->left=NULL;
		while(i){
			free(i->name);
			free(i->id);
			XDestroyImage(i->image);
			prev=i;
			i=i->left;
			free(prev);
		}
	}

	XFreeGC(g.x.display,g.x.gc);
	XFreeGC(g.x.display,g.x.rgc);
	
	XFreePixmap(g.x.display,g.x.buffer);
	XCloseDisplay(g.x.display);
	return;
}

